### PR TITLE
🐛 KCP should update Kubeadm config map when setting `imageRepository`

### LINKED ISSUE #2805: KCP upgrades with imageRepository changed aren't complete
**What steps did you take and what happened:**

when performing a KCP upgrade from `v1.17.1` to `v1.17.3` and updating KCP's `imageRepository` in the `clusterConfiguration`, `dns` and `etcd` the KCP is edited successfully. but when reconciling the upgrade kubeadm is trying to pull the core k8s components from the old `imageRepository` (coreDNS and kube-proxy didn't get updated too). 

when checking the kubeadm configmap the `imageRepository` for etcd is the only one updated, the other ones are still pointing to the old registry.

cc @ncdc @vincepri @randomvariable @benmoss 

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug

### PR DESCRIPTION
**Pull request by benmoss**
**What this PR does / why we need it**:
Fixes issues arising when a user specifies a global `imageRegistry` in their `clusterConfiguration`. We weren't updating the kubeadm configmap and so when we spun up new nodes kubeadm would still pull from the default registry.

**Which issue(s) this PR fixes**
Fixes #2805

### CODE REVIEW

**Comment by vincepri** [comment]
/retitle :bug: KCP should update Kubeadm config map when setting `imageRepository`

**Comment by benmoss** [comment]
@k8s-ci-robot hates emojis

File: controlplane/kubeadm/controllers/upgrade.go [hunk 1]
```diff
@@ -64,6 +64,13 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, errors.Wrap(err, "failed to update the kubernetes version in the kubeadm config map")
 	}
 
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration != nil {
+		imageRepository := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository
+		if err := workloadCluster.UpdateImageRepositoryInKubeadmConfigMap(ctx, imageRepository); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to update the kubernetes version in the kubeadm config map")
```
Comment (benmoss) [inline controlplane/kubeadm/controllers/upgrade.go]: This is a bad err message!

File: controlplane/kubeadm/internal/kubeadm_config_map.go [hunk 2]
```diff
@@ -89,6 +90,30 @@ func (k *kubeadmConfig) UpdateKubernetesVersion(version string) error {
...
+	data, ok := k.ConfigMap.Data[clusterConfigurationKey]
+	if !ok {
+		return errors.Errorf("unable to find %q key in kubeadm ConfigMap", clusterConfigurationKey)
+	}
+	configuration, err := yamlToUnstructured([]byte(data))
+	if err != nil {
+		return errors.Wrapf(err, "unable to decode kubeadm ConfigMap's %q to Unstructured object", clusterConfigurationKey)
+	}
+	if err := unstructured.SetNestedField(configuration.UnstructuredContent(), imageRepository, configImageRepositoryKey); err != nil {
+		return errors.Wrapf(err, "unable to update %q on kubeadm ConfigMap's %q", configVersionKey, clusterConfigurationKey)
```
Comment (vincepri) [inline controlplane/kubeadm/internal/kubeadm_config_map.go]:
```suggestion
		return errors.Wrapf(err, "unable to update %q on kubeadm ConfigMap's %q", imageRepository, clusterConfigurationKey)
```

**Pull request merged**